Sources/MongoSwift/MongoCollection+BulkWrite.swift

```diff
@@ -17,118 +17,385 @@ extension MongoCollection {
-    public let update: Document
+    /// Adds the `insertOne` operation to a bulk write and returns its `_id`
+    fileprivate func addToBulkWrite(bulk: OpaquePointer) throws -> BsonValue? {
+        let encoder = BsonEncoder()
 
-    /// A set of filters specifying to which array elements an update should apply.
-    public let arrayFilters: [Document]?
+        let document = try encoder.encode(self.document)
+        // TODO: Differentiate between unset _id and explicit null (SWIFT-162)
+        if document["_id"] == nil {
+            try ObjectId().encode(to: document.storage, forKey: "_id")
+        }
 
-    /// Specifies a collation to use.
-    public let collation: Document?
+        var error = bson_error_t()
 
-    /// When `true`, creates a new document if no document matches the query.
+        if !mongoc_bulk_operation_insert_with_opts(bulk, document.data, nil, &error) {
+            throw MongoError.invalidArgument(message: toErrorString(error))
+        }
+
+        return document["_id"]
+    }
+}
+
+internal struct ReplaceOneModelOptions: Encodable {
+    public let collation: Document?
     public let upsert: Bool?
 }
 
-/// A model for an `updateMany` command.
-public struct UpdateManyModel: WriteModel {
-    /// A `Document` representing the match criteria.
-    public let filter: Document
+/// A model for a `replaceOne` operation within a bulk write.
+public struct ReplaceOneModel<T: Codable>: DeleteOrUpdateWriteModel {
+    private let filter: Document
+    private let replacement: T
+    private let options: ReplaceOneModelOptions
 
-    /// A `Document` containing update operators.
-    public let update: Document
+    /**
+     * Create a `replaceOne` operation for a bulk write.
+     *
+     * - Parameters:
+     *   - filter: A `Document` representing the match criteria
+     *   - replacement: The `Codable` (e.g. `Document`) to use as the replacement value
+     *   - collation: Specifies a collation to use
+     *   - upsert: When `true`, creates a new document if no document matches the query
+     */
+    public init(filter: Document, replacement: T, collation: Document? = nil, upsert: Bool? = nil) {
+        self.filter = filter
+        self.replacement = replacement
+        self.options = ReplaceOneModelOptions(collation: collation, upsert: upsert)
+    }
 
-    /// A set of filters specifying to which array elements an update should apply.
-    public let arrayFilters: [Document]?
+    /// Adds the `replaceOne` operation to a bulk write
+    fileprivate func addToBulkWrite(bulk: OpaquePointer) throws {
+        let encoder = BsonEncoder()
+        let replacement = try encoder.encode(self.replacement)
+        let opts = try encoder.encode(self.options)
+        var error = bson_error_t()
 
-    /// Specifies a collation to use.
-    public let collation: Document?
+        if !mongoc_bulk_operation_replace_one_with_opts(bulk, filter.data, replacement.data, opts.data, &error) {
```

Comment:
I know this is consistent with the rest of the codebase, but this just reminded me: I've been thinking lately that in these cases it might be more Swift-y to do `guard mongoc_some_operation(...) else { // throw err }` rather than `if !mongoc_some_operation(...) { // throw err }`. 
it kind of seems to me like `guard` is appropriate when we expect something to work, which we do in all these cases. vs we should use `if` when it could go either way. what do you think?
(perhaps something for us to address in a future ticket rather than here)

Reply:
I agree that this seems like a fine use of `guard`. More generally, any negated if condition and a return is probably a code smell that we should utilize `guard` instead. Happy to incorporate this change in this PR and I've created [SWIFT-201](https://jira.mongodb.org/browse/SWIFT-201) to track the refactor elsewhere.